src/field/FieldEquation.i.hh

```diff
@@ -0,0 +1,73 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2020 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file FieldEquation.i.cuh
+//---------------------------------------------------------------------------//
+#include "MagField.hh"
+#include <cmath>
+#include "base/Constants.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with a magnetic field.
+ */
+CELER_FUNCTION
+FieldEquation::FieldEquation(MagField& field) : field_(field)
+{
+    // set the default charge and the lorentz_cof with -eplus
+    set_charge(-1.0);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Set charge for this track and recalcuate lorentz_cof
+ */
+
+CELER_FUNCTION void FieldEquation::set_charge(real_type charge)
+{
+    charge_ = charge;
+    coeffi_ = charge_ * constants::c_light * 1.0e-8;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Evaluate the right hand side of the field equation
+ */
+
+CELER_FUNCTION void FieldEquation::
+                    operator()(const ode_type y, ode_type& dydx) const
+{
+    this->evaluate_rhs(field_(), y, dydx);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Evaluate the right hand side of the Lorentz equation for a given B
+ */
+
+CELER_FUNCTION
+void FieldEquation::evaluate_rhs(const Real3    B,
+                                 const ode_type y,
+                                 ode_type&      dy) const
+{
+    /*** m d^2x/dt^2 = (q/c)(vxB), s = |v|t, y = dx/ds
+         rhs:  dx/ds = v/|v|
+               dy/ds = (q/pc)(yxB)
+    */
+
+    dy[0] = y[3];
+    dy[1] = y[4];
+    dy[2] = y[5];
+    dy[3] = coeffi_ * (y[4] * B[2] - y[5] * B[1]);
+    dy[4] = coeffi_ * (y[5] * B[0] - y[3] * B[2]);
+    dy[5] = coeffi_ * (y[3] * B[1] - y[4] * B[0]);
```

Comment:
These were my original expressions too, but realized that pos and mon are Real3 (so, op= does not work here).  I also have not introduced modifiers for OdeArray position and momentum yet.  I am going to use vecgeom::Vector3D for three vector in all field codes in the next round of commits (if Real3 will be a just plain array as now) and use the celeritas unit consistently every where - current unit conventions are mm for pos, MeV for mon and field value in Tesla which were assumed during initialization of FieldTrackView and MagField.  The same story for the c*1e-8 which will be consistently changed by dropping the clhep-style unit conventions later.  Of course two equations do not make sense without the last line.

Reply:
`op=` should definitely work for `Real3` -- otherwise we wouldn't be able to set the outgoing directions of secondaries, among other things.

Please use the celeritas data structures inside celeritas code, not the vecgeom ones. If there are features such as `operator*=` that you feel are critical to getting this online, we can add them to the array definition, or (since Real3 has no private data) you can implement free functions in a header file inside the `field` directory.

Reply:
Okay.  May I add a scale function, something like
    template<class T, std::size_t N>                                            
    CELER_FUNCTION Array<T, N> scale_array(const T a, const Array<T, N>& x)           
    {                                                                           
        Array<T, N> result;                                                     
        for (std::size_t i = 0; i != N; ++i)                                    
        {                                                                       
            result[i] = a * x[i];                                               
        }                                                                       
         return result;                                                         
    }                        

Otherwise, I will use a private function (simplified one equivalent to scale_array<real_type, 3>) inside the class (then it will be just add one more function call for nothing).

Reply:
Sure thing. If it's a const transformation like this I would call it `scaled` (indicating it returns a scaled version) rather than `scale` (indicating it modifies the original).